Fix for HistogramObserver

TiRune opened the pull request:
Summary:
There were two problems with the HistogramObserver:
1. It does not work when someone passes a batch_size 1, tensor_size 1 data-point.
2. The Histogram doesn't seem to actually update if the range of the new x falls within the old one

These issues were both fixed.

On top of this, I greatly simplified the logic for the histogram updating. Now, it doesn't do the downsampling anymore, which saves a ton of memory and code. The accuracy can still be controlled with the upsampling ratio. This ratio was also too high for the accuracy we generally need here, I reduced the default for this.
Also the code is cleaner now, much easier to follow what's happening.

test_histogram_observer_same_inputs was likely wrong - If I pass 0s and 1s to my histogramobserver, I want them to actually count! The current test now thinks it's good to discard and ignore these values.

Test Plan: You can run the included tests.

Differential Revision: D58931336

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1131,66 +1131,74 @@ def _non_linear_param_search(self) -> Tuple[torch.Tensor, torch.Tensor]:
+        update_min: torch.Tensor,
+        update_max: torch.Tensor,
+        n_bins: int,
+        upsample_rate: int = 4,
```
why is this not the same with L967?

TiRune replied:
Good point. The original code was written in this way, essentially like a static_method. 
Do you think it's better if I rewrote everything to use just self. everywhere? Or do you like the more explicit static_method kind of setup? 

I removed the default from the static_method - but calling with self.upsample_rate in _combine_histograms. If you want me to go full ham, I can remove most passing and call e.g. self. histogram everywhere instead of passing it.

jerryzh168 replied:
yeah it seems that we don't really need to make this a staticmethod, I feel we can just do `self` everywhere and just remove the args

TiRune replied:
Fixed this and made things non-static

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1131,66 +1131,74 @@ def _non_linear_param_search(self) -> Tuple[torch.Tensor, torch.Tensor]:
+        histogram = histogram.repeat_interleave(upsample_rate) / upsample_rate
+        bin_size = (orig_max - orig_min) / (n_bins * upsample_rate)
+        mid_points_histogram = (
+            torch.linspace(orig_min, orig_max, n_bins * upsample_rate + 1)[:-1]
+            + 0.5 * bin_size
+        )
+        new_bin_size = (update_min - update_max) / n_bins
+        boundaries_new_histogram = torch.linspace(update_min, update_max, n_bins)
+        # this maps the mid-poits of the histogram to the new histogram's space
+        bucket_assignments = (
+            torch.bucketize(mid_points_histogram, boundaries_new_histogram) - 1
         )
-        e = downsample_rate / upsample_rate * (self.max_val - self.min_val) - (combined_max - combined_min)
-        start_idx = int(
-            torch.round((self.min_val - combined_min) / (self.max_val - self.min_val) * self.bins * upsample_rate).item()
+        # this then maps the histogram mid-points in the new space, weighted by the original histogram's values
+        # this is just the old histogram in the new histogram's space
+
+        update_histogram = torch.bincount(
+            bucket_assignments, weights=histogram, minlength=n_bins
         )
```
this is great! it's really nice that we can use all these existing torch functions to implement this, many of these I saw the first time and yet it all make sense after I look up the docs :)

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1204,7 +1212,8 @@ def reset_histogram(self, x: torch.Tensor, min_val: torch.Tensor, max_val: torch
         )
 
-    def forward(self, x_orig: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
nit: maybe just keep the old name, L1219 also reassigns x_orig.detach() to x again, seems a bit confusing to assign between x_orig and x back and forth

TiRune replied:
Changing it back then - VScode was complaining the signature doesn't match up with the base class!

TiRune replied:
I just added a pyre-ignore and kept the interface the same. LMK if you prefer the renaming instead.

jerryzh168 reviewed (approved):
thanks a lot! I think the implementation makes sense, we can land as long as we pass the CI and also make sure this doesn't have the OOM issue

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1204,7 +1212,8 @@ def reset_histogram(self, x: torch.Tensor, min_val: torch.Tensor, max_val: torch
+        x_orig = x
         if x_orig.numel() == 0:
             return x_orig
         x = x_orig.detach()
```
we can rename this to `x_detach` to reduce confusion I think

TiRune replied:
This makes sense if we keep 'x' as the input. It does clutter the code a bit. If we decide on having 'x' as input instead of 'x_orig' I would make that change, otherwise keep it as is.

jerryzh168 commented on torch/ao/quantization/observer.py:
```diff
@@ -1204,7 +1212,8 @@ def reset_histogram(self, x: torch.Tensor, min_val: torch.Tensor, max_val: torch
-    def forward(self, x_orig: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x_orig = x
         if x_orig.numel() == 0:
```
and also rename `x_orig` to `x`

TiRune commented:
@jerryzh168 It doesn't really look as if those failing trunk tests are related to this diff right? What do I do about these? :D

jerryzh168 commented:
> @jerryzh168 It doesn't really look as if those failing trunk tests are related to this diff right? What do I do about these? :D

yeah doesn't look related, we can rebase after hud turns green: https://hud.pytorch.org/ or we could also force land if this is urgent (with `@pytorchbot merge -f {MESSAGE}` command)

TiRune commented:
@pytorchbot merge

TiRune committed:
Fix for HistogramObserver (#129387)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/129387

There were two problems with the HistogramObserver:
1. It does not work when someone passes a batch_size 1, tensor_size 1 data-point.
2. The Histogram doesn't seem to actually update if the range of the new x falls within the old one (https://www.internalfb.com/code/fbsource/[8f6ad6596b6d385ed968da81e2e3e42a219c59ea]/fbcode/caffe2/torch/ao/quantization/observer.py?lines=1255-1256)

These issues were both fixed.

On top of this, I greatly simplified the logic for the histogram updating. Now, it doesn't do the downsampling anymore, which saves a ton of memory and code. The accuracy can still be controlled with the upsampling ratio. This ratio was also too high for the accuracy we generally need here, I reduced the default for this.
Also the code is cleaner now, much easier to follow what's happening.

Test Plan:
N5523589

You can test the implementation here in an easy way.

You can also run the tests, added a test for both scenarios that went wrong before.

Reviewed By: balakv504

Differential Revision: D58931336

TiRune commented:
@pytorchbot merge

Xia-Weiwen commented:
This PR causes backward compatibility issue. The `upsample_rate` is removed.😒

TiRune commented:
> This PR causes backward compatibility issue. The `upsample_rate` is removed.😒

We decided to remove it - since it’s not likely a parameter a user would care about. 
Is there a reason you want it back? Are you actively using it in some way? 😄

Xia-Weiwen commented:
> > This PR causes backward compatibility issue. The `upsample_rate` is removed.😒
> 
> We decided to remove it - since it’s not likely a parameter a user would care about. Is there a reason you want it back? Are you actively using it in some way? 😄

We serialize observers by saving their parameters in a JSON file. Then we reconstruct observers by loading these parameters and call the constructor. Now the `upsample_rate` is removed, old JSON files cannot be used because they have `upsample_rate`. Maybe there is no perfect solution. If you add `upsample_rate` back, it has to be there forever for compatibility. I guess we will have to fix it on our side.

iperminov-mythic commented:
The recently released PyTorch 2.4.0 includes a commit that introduced a bug fixed by this PR, but does not include the fix :-(
